src/build_system.ml

```diff
@@ -183,6 +131,55 @@ module Internal_rule = struct
        [if_file_exists] are resolved inside the [Build.t] value. *)
     ignore (Lazy.force t.static_deps : Build_interpret.Static_deps.t);
     Build_interpret.lib_deps t.build
+
+  let root =
+    { id          = Id.gen ()
+    ; static_deps = lazy { rule_deps = Path.Set.empty
+                         ; action_deps = Path.Set.empty
+                         }
+    ; targets     = Path.Set.empty
+    ; context     = None
+    ; build       = Build.return (Action.Progn [])
+    ; mode        = Standard
+    ; loc         = None
+    ; dir         = Path.root
+    ; exec        = Not_started { eval_rule = (fun _ -> assert false)
+                                ; exec_rule = (fun _ -> assert false)
+                                }
+    ; rev_deps    = []
+    ; all_targets = Path.Set.empty
+    }
+
+  let dependency_cycle last t =
+    let rec build_loop acc t =
+      if Path.Set.mem t.targets last then
+        last :: acc
+      else
+        match
+          List.find t.rev_deps ~f:(fun (_, t) ->
+            Path.Set.mem t.all_targets last)
+        with
+        | None -> assert false
```

Comment:
We can use `Option.value_exn` here (there's a couple `List.find` uses like this in the code base, so extracting a `List.find_exn` is a nice alternative as well)

Reply:
Indeed. I'm using `Option.value_exn` for now, we can add `List.find_exn` and change all the occurrences at once in another PR

Reply:
OK I'll take care of this part.